DESIGN.md

```diff
@@ -0,0 +1,155 @@
+
+The field `accept` is set to `true` in order to signal that any subsequent input, including the lack of any remaining bytes, is acceptable under the given format, and that no further processing is necessary to disambiguate possible alternatives. If `accept` is `true`, it is to be expected that `branches` is empty; in the case that `accept` is false and `branches` is empty, it is to be understood that the committed choice of prefix has no valid interpretation within the original format, regardless of what input follows. This is a special case of the general rule that, if `accept` is `false`, any byte-values that are not members of any byte-set in `branches`, or end-of-input, would both mark the current input as a non-match for the chosen descent that led to a the current `MatchTreeStep`.
+
+### MatchTreeLevel
+
+Concisely, a `MatchTreeLevel` represents the unified branching conditions of all
+choice-points at a consistent lookahead distance. That is, in a tree diagram
+with a constant vertical height for all edges between nodes and their children,
+a `MatchTreeLevel` can be thought of as a stratum found at a fixed depth.
+
+Though subject to change, it is currently defined as:
+
+```rust
+#[derive(Clone, Debug)]
+struct MatchTreeLevel<'a> {
+    accept: Option<usize>,
+    branches: Vec<(ByteSet, LevelBranch<'a>)>,
+}
+```
+
+where the `accept` field has the same semantics the like-named field in `MatchTree`.
+
+The `branches` field is defined in terms of a type-alias `LevelBranch<'_>`,
+outlined in the subsection found below
+
+#### LevelBranch
+
+The type-alias `LevelBranch` is defined as follows:
+
+```rust
+type LevelBranch<'a> = HashSet<(usize, Rc<Next<'a>>)>;
+```
+
+To understand the significance of this type, we make the following observation:
+
+In order for a `MatchTree` to be well-formed, it and all of its descendants
+must have mutually disjoint branching-predicates. That is, for all well-formed `tree : MatchTree`, it must be the case that for all values of `b : u8`, the iterator generated by `tree.branches.filter(|(s, _)| s.contains(b)))` must yield zero or one elements only.
+
+However, the same does not hold for `MatchTreeLevels`, as they represent the superposition of all choice-points at a common depth. If that depth is non-zero, the `MatchTree` may still be well-formed even if two choice-points within the `MatchTreeLevel` have some non-null intersection among the cartesian product of their respective lists of branching-predicates.
+
+For this reason, a `MatchTreeLevel` cannot represent a single, unambiguous descent into a follow-set for any given byte of input, but rather, must form a disjunction over the smallest set of mutually disjoint subsets of all the branching predicates of its constituent choice-points, such as in the following algorithm (in pseud-ocode):
```

Comment:
Typo, pseudo-code.

Reply:
Ah, indeed. I left out the hyphen initially and clearly didn't put it in the right place :sweat_smile: